File: actualizar.js
const { MongoClient } = require('mongodb');
const inquirer = require('inquirer');

// Conexión a MongoDB (Actualiza con tu URI)
const mongoURI = 'mongodb+srv://dia21066:CMSOw7YTLnPTsdEp@cluster0.a4fng.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0';

const connectToMongoDB = async () => {
  const client = new MongoClient(mongoURI, { useNewUrlParser: true, useUnifiedTopology: true });
  await client.connect();
  console.log("Conectado a MongoDB");
  return client;
};

// Actualizar Usuario
const actualizarUsuario = async () => {
  const { nombre, nuevoCorreo, nuevaDireccion, nuevoTelefono } = await inquirer.prompt([
    { type: 'input', name: 'nombre', message: 'Ingrese el nombre del usuario a actualizar:' },
    { type: 'input', name: 'nuevoCorreo', message: 'Ingrese el nuevo correo del usuario:' },
    { type: 'input', name: 'nuevaDireccion', message: 'Ingrese la nueva dirección del usuario:' },
    { type: 'input', name: 'nuevoTelefono', message: 'Ingrese el nuevo teléfono del usuario:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const usuariosCollection = db.collection('Usuarios');

  const result = await usuariosCollection.updateOne(
    { nombre },
    { $set: { correo: nuevoCorreo, direccion: nuevaDireccion, telefono: nuevoTelefono } }
  );

  if (result.modifiedCount > 0) {
    console.log('Usuario actualizado con éxito');
  } else {
    console.log('No se encontró el usuario o no hubo cambios');
  }
  await client.close();
};

// Actualizar Restaurante
const actualizarRestaurante = async () => {
  const { nombre, nuevaDireccion, nuevaCategoria } = await inquirer.prompt([
    { type: 'input', name: 'nombre', message: 'Ingrese el nombre del restaurante a actualizar:' },
    { type: 'input', name: 'nuevaDireccion', message: 'Ingrese la nueva dirección del restaurante:' },
    { type: 'input', name: 'nuevaCategoria', message: 'Ingrese la nueva categoría del restaurante:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const restaurantesCollection = db.collection('Restaurantes');

  const result = await restaurantesCollection.updateOne(
    { nombre },
    { $set: { direccion: nuevaDireccion, categoria: nuevaCategoria } }
  );

  if (result.modifiedCount > 0) {
    console.log('Restaurante actualizado con éxito');
  } else {
    console.log('No se encontró el restaurante o no hubo cambios');
  }
  await client.close();
};

// Actualizar Artículo de Menú
const actualizarArticuloDeMenu = async () => {
  const { nombre, nuevaDescripcion, nuevoPrecio, nuevaCategoria } = await inquirer.prompt([
    { type: 'input', name: 'nombre', message: 'Ingrese el nombre del artículo a actualizar:' },
    { type: 'input', name: 'nuevaDescripcion', message: 'Ingrese la nueva descripción del artículo:' },
    { type: 'input', name: 'nuevoPrecio', message: 'Ingrese el nuevo precio del artículo:' },
    { type: 'input', name: 'nuevaCategoria', message: 'Ingrese la nueva categoría del artículo:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const articulosCollection = db.collection('Articulos_Menu');

  const result = await articulosCollection.updateOne(
    { nombre },
    { $set: { descripcion: nuevaDescripcion, precio: parseFloat(nuevoPrecio), categoria: nuevaCategoria } }
  );

  if (result.modifiedCount > 0) {
    console.log('Artículo de menú actualizado con éxito');
  } else {
    console.log('No se encontró el artículo o no hubo cambios');
  }
  await client.close();
};

// Actualizar Orden
const actualizarOrden = async () => {
  const { idOrden, nuevoEstado } = await inquirer.prompt([
    { type: 'input', name: 'idOrden', message: 'Ingrese el ID de la orden a actualizar:' },
    { type: 'input', name: 'nuevoEstado', message: 'Ingrese el nuevo estado de la orden (Ejemplo: Pendiente, Enviado, Entregado):' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const ordenesCollection = db.collection('Ordenes');

  const result = await ordenesCollection.updateOne(
    { _id: new MongoClient.ObjectId(idOrden) },
    { $set: { estado: nuevoEstado } }
  );

  if (result.modifiedCount > 0) {
    console.log('Orden actualizada con éxito');
  } else {
    console.log('No se encontró la orden o no hubo cambios');
  }
  await client.close();
};

// Actualizar Reseña
const actualizarResena = async () => {
  const { idResena, nuevaCalificacion, nuevoComentario } = await inquirer.prompt([
    { type: 'input', name: 'idResena', message: 'Ingrese el ID de la reseña a actualizar:' },
    { type: 'input', name: 'nuevaCalificacion', message: 'Ingrese la nueva calificación (1-5):' },
    { type: 'input', name: 'nuevoComentario', message: 'Ingrese el nuevo comentario:' },
  ]);

  const client = await connectToMongoDB();
  const db = client.db('Proyecto_2');
  const resenasCollection = db.collection('Resenas');

  const result = await resenasCollection.updateOne(
    { _id: new MongoClient.ObjectId(idResena) },
    { $set: { calificacion: parseFloat(nuevaCalificacion), comentario: nuevoComentario } }
  );

  if (result.modifiedCount > 0) {
    console.log('Reseña actualizada con éxito');
  } else {
    console.log('No se encontró la reseña o no hubo cambios');
  }
  await client.close();
};

//exportar las funciones
module.exports = {
  actualizarUsuario,
  actualizarRestaurante,
  actualizarArticuloDeMenu,
  actualizarOrden,
  actualizarResena,
};

File: actualizar.py
from flask import jsonify, request
from pymongo import MongoClient
from bson import ObjectId

# Conexión a MongoDB
def connect_to_mongo():
    mongo_uri = 'mongodb+srv://dia21066:CMSOw7YTLnPTsdEp@cluster0.a4fng.mongodb.net/?retryWrites=true&w=majority'
    client = MongoClient(mongo_uri)
    db = client['Proyecto_2'] 
    return db

def _filtrar_campos(data):
    return {
        key: value
        for key, value in data.items()
        if value is not None and not (isinstance(value, str) and value.strip() == "")
    }


def _armar_filtro(id_str):
    # Si es hex válido, considerar ObjectId y string
    if ObjectId.is_valid(id_str):
        return {"$or": [{"_id": ObjectId(id_str)}, {"_id": id_str}]}
    return {"_id": id_str}


def actualizar_usuario(request, db):
    try:
        id_str = request.view_args['id']
        filtro = _armar_filtro(id_str)
        data = request.get_json() or {}
        update_data = _filtrar_campos(data)
        if not update_data:
            return jsonify({"message": "Sin campos para actualizar"}), 400

        result = db.Usuarios.update_one(filtro, {"$set": update_data})
        if result.matched_count == 0:
            return jsonify({"message": "Usuario no encontrado"}), 404
        if result.modified_count == 0:
            return jsonify({"message": "No se encontraron cambios para actualizar"}), 200
        return jsonify({"message": "Usuario actualizado con éxito"}), 200

    except Exception as e:
        return jsonify({"message": f"Error al actualizar el usuario: {str(e)}"}), 500


def actualizar_restaurante(request, db):
    try:
        id_str = request.view_args['id']
        filtro = _armar_filtro(id_str)
        data = request.get_json() or {}
        campos = _filtrar_campos(data)
        set_data = {}

        # Dirección anidada
        if "calle" in campos:
            set_data["direccion.calle"] = campos.pop("calle")
        if "ciudad" in campos:
            set_data["direccion.ciudad"] = campos.pop("ciudad")
        # Coordenadas
        lon = campos.pop("longitud", None)
        lat = campos.pop("latitud", None)
        if lon is not None and lat is not None:
            set_data["ubicacion.coordinates"] = [lon, lat]
        # Resto campos planos
        for k, v in campos.items():
            set_data[k] = v

        if not set_data:
            return jsonify({"message": "Sin campos para actualizar"}), 400

        result = db.Restaurantes.update_one(filtro, {"$set": set_data})
        if result.matched_count == 0:
            return jsonify({"message": "Restaurante no encontrado"}), 404
        if result.modified_count == 0:
            return jsonify({"message": "No se encontraron cambios para actualizar"}), 200
        return jsonify({"message": "Restaurante actualizado con éxito"}), 200

    except Exception as e:
        return jsonify({"message": f"Error al actualizar el restaurante: {str(e)}"}), 500


def actualizar_resena(request, db):
    try:
        id_str = request.view_args['id']
        filtro = _armar_filtro(id_str)
        data = request.get_json() or {}
        campos = _filtrar_campos(data)
        # Mapear claves de relación
        if "usuarioIdResena" in campos:
            campos["usuarioId"] = campos.pop("usuarioIdResena")
        if "restauranteIdResena" in campos:
            campos["restauranteId"] = campos.pop("restauranteIdResena")
        if not campos:
            return jsonify({"message": "Sin campos para actualizar"}), 400

        result = db.Resenas.update_one(filtro, {"$set": campos})
        if result.matched_count == 0:
            return jsonify({"message": "Reseña no encontrada"}), 404
        if result.modified_count == 0:
            return jsonify({"message": "No se encontraron cambios para actualizar"}), 200
        return jsonify({"message": "Reseña actualizada con éxito"}), 200

    except Exception as e:
        return jsonify({"message": f"Error al actualizar la reseña: {str(e)}"}), 500


def actualizar_articulo_de_menu(request, db):
    try:
        id_str = request.view_args['id']
        filtro = _armar_filtro(id_str)
        data = request.get_json() or {}
        update_data = _filtrar_campos(data)
        if not update_data:
            return jsonify({"message": "Sin campos para actualizar"}), 400

        result = db.Articulos_Menu.update_one(filtro, {"$set": update_data})
        if result.matched_count == 0:
            return jsonify({"message": "Artículo no encontrado"}), 404
        if result.modified_count == 0:
            return jsonify({"message": "No se encontraron cambios para actualizar"}), 200
        return jsonify({"message": "Artículo de menú actualizado con éxito"}), 200

    except Exception as e:
        return jsonify({"message": f"Error al actualizar el artículo de menú: {str(e)}"}), 500


def actualizar_orden(request, db):
    try:
        id_str = request.view_args['id']
        filtro = _armar_filtro(id_str)
        data = request.get_json() or {}
        campos = _filtrar_campos(data)
        # Mapear y convertir relaciones
        if "usuarioId" in campos:
            campos["usuarioId"] = campos["usuarioId"]
        if "restauranteId" in campos:
            campos["restauranteId"] = campos["restauranteId"]
        if "items" in campos and isinstance(campos["items"], list):
            campos["items"] = campos["items"]
        if not campos:
            return jsonify({"message": "Sin campos para actualizar"}), 400

        result = db.Ordenes.update_one(filtro, {"$set": campos})
        if result.matched_count == 0:
            return jsonify({"message": "Orden no encontrada"}), 404
        if result.modified_count == 0:
            return jsonify({"message": "No se encontraron cambios para actualizar"}), 200
        return jsonify({"message": "Orden actualizada con éxito"}), 200

    except Exception as e:
        return jsonify({"message": f"Error al actualizar la orden: {str(e)}"}), 500
